crates/esthri/src/bufferpool.rs

```diff
@@ -0,0 +1,96 @@
+use std::{iter, sync::Arc};
+
+use bytes::{Bytes, BytesMut};
+use once_cell::sync::OnceCell;
+use tokio::sync::{Mutex, MutexGuard, Notify};
+
+use crate::Config;
+
+static POOL: OnceCell<Pool> = OnceCell::new();
+
+pub struct Pool {
+    shared: Arc<Shared>,
+}
+
+impl Pool {
+    pub fn global() -> &'static Pool {
```

Comment:
Isn't the disadvantage of pooling like this is that once we've allocated a pool in the buffer, esthri will hold that memory until it closes? I guess this is probably OK for the CLI but as a library it seems like it could be problematic because an application might upload (causing the esthri to grab up to `concurrent_upload_tasks*upload_part_size` memory) and then do other non-esthri things afterwards. And when it's doing the non-esthri things afterwards, the OS won't be able to claim back that memory that esthri has reserved? In practice, I guess this is probably OK for the SITL worker at the moment, as it does all its uploading at the end and then exits, but it seems like it might not be the best practice (maybe we'd want to daemonify the worker at some stage which wouldn't play well with this, for eg).

Maybe I'm missing something but what benefits does having our own pool give? Quicker allocations? AFAICT, the memory use of esthri is already kind of bounded by `concurrent_upload_tasks` and `upload_part_size` so I'm not sure that this pool gives us much benefit?

Reply:
Oh yeah we definitely should not keep this, on top of what you said this is just broken. Once I heard you had success with the uploading I just gave up on it. The downloading I have here doesn't explicitly allocate anything so doesn't use the pool

Reply:
I think the main things worth taking from this are 

- the implementation of download impl
- the PendingUploads tracking/ctrl-c handling 
- replacement of a bunch of blocking api calls

Everything else is less of an improvement and just style kind of stuff/minor things